src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala

```diff
@@ -537,35 +540,60 @@ trait BCodeBodyBuilder extends BCodeSkelBuilder {
       else assert(!int.shouldEmitJumpAfterLabels) // scalac
     }
 
-    private def genReturn(r: Return) = r match{
-      case Return(expr) =>
-      val returnedKind = tpeTK(expr)
-      genLoad(expr, returnedKind)
-      adapt(returnedKind, returnType)
-      val saveReturnValue = (returnType != UNIT)
-      lineNumber(r)
+    private def genLabeled(tree: Labeled): BType = tree match {
+      case Labeled(bind, expr) =>
 
-      cleanups match {
-        case Nil =>
-          // not an assertion: !shouldEmitCleanup (at least not yet, pendingCleanups() may still have to run, and reset `shouldEmitCleanup`.
-          bc emitRETURN returnType
-        case nextCleanup :: rest =>
-          if (saveReturnValue) {
-            if (insideCleanupBlock) {
-              int.warning(r.pos, "Return statement found in finally-clause, discarding its return-value in favor of that of a more deeply nested return.")
-              bc drop returnType
-            } else {
-              // regarding return value, the protocol is: in place of a `return-stmt`, a sequence of `adapt, store, jump` are inserted.
-              if (earlyReturnVar == null) {
-                earlyReturnVar = locals.makeLocal(returnType, "earlyReturnVar", expr.tpe, expr.pos)
+      val resKind = tpeTK(tree)
+      genLoad(expr, resKind)
+      markProgramPoint(programPoint(bind.symbol))
+      resKind
+    }
+
+    private def genReturn(r: Return): Unit = r match {
+      case Return(expr, fromSym) =>
+
+      if (NoSymbol == fromSym) {
```

Comment:
The indentation is weird here, normally we add a level after `case ... =>`

Reply:
I agree, but look around: all `genXYZ` methods have this weird format to extract their node. I followed the local style.